workflow/controller/controller.go

```diff
@@ -542,6 +541,52 @@ func (wfc *WorkflowController) runPodCleanup(ctx context.Context) {
 	}
 }
 
+func (wfc *WorkflowController) getPodFinalizerRemovalPatch(namespace, podName string) (string, bool) {
```

Comment:
Wouldn't it be better to generate a JSON merge patch from the original and modified objects, rather than creating one manually? For example, something like this:

```diff
diff --git a/workflow/controller/controller.go b/workflow/controller/controller.go
index bb0b84d8a..f62649465 100644
--- a/workflow/controller/controller.go
+++ b/workflow/controller/controller.go
@@ -10,6 +10,7 @@ import (
        "syscall"
        "time"

+       jsonpatch "github.com/evanphx/json-patch"
        "github.com/upper/db/v4"

        "github.com/argoproj/pkg/errors"
@@ -541,24 +542,38 @@ func (wfc *WorkflowController) runPodCleanup(ctx context.Context) {
        }
 }

-func (wfc *WorkflowController) getPodFinalizerRemovalPatch(namespace, podName string) (string, bool) {
+func (wfc *WorkflowController) getPodFinalizerRemovalPatch(namespace, podName string) ([]byte, error) {
        if os.Getenv("ARGO_POD_STATUS_CAPTURE_FINALIZER") != "true" {
-               return "", false
+               return nil, nil
        }
-       found := false
+       oldPod, err := wfc.getPod(namespace, podName)
+       if err != nil {
+               return nil, nil
+       }
+
+       newPod := oldPod.DeepCopy()
        keeps := make([]string, 0)
-       pod, err := wfc.getPod(namespace, podName)
-       if err == nil && pod != nil {
-               for _, f := range pod.ObjectMeta.Finalizers {
-                       if f == common.FinalizerPodStatus {
-                               found = true
-                       } else {
-                               keeps = append(keeps, f)
-                       }
+       for _, f := range newPod.ObjectMeta.Finalizers {
+               if f != common.FinalizerPodStatus {
+                       keeps = append(keeps, f)
                }
        }
-       finalizers, _ := json.Marshal(keeps)
-       return string(finalizers), found
+       newPod.Finalizers = keeps
+
+       oldData, err := json.Marshal(oldPod)
+       if err != nil {
+               return nil, err
+       }
+       newData, err := json.Marshal(newPod)
+       if err != nil {
+               return nil, err
+       }
+
+       mergePatch, err := jsonpatch.CreateMergePatch(oldData, newData)
+       if err != nil {
+               return nil, err
+       }
+       return mergePatch, nil
 }

 func (wfc *WorkflowController) podHasFinalizer(pod *apiv1.Pod) bool {
@@ -571,18 +586,23 @@ func (wfc *WorkflowController) podHasFinalizer(pod *apiv1.Pod) bool {
 }

 func (wfc *WorkflowController) removePodFinalizer(ctx context.Context, namespace, podName string) error {
-       if finalizers, found := wfc.getPodFinalizerRemovalPatch(namespace, podName); found {
-               // merge patch is idempotent
-               _, err := wfc.kubeclientset.CoreV1().Pods(namespace).Patch(
-                       ctx,
-                       podName,
-                       types.MergePatchType,
-                       []byte(fmt.Sprintf(`{"metadata": {"finalizers": %s}}`, finalizers)),
-                       metav1.PatchOptions{},
-               )
-               if err != nil {
-                       return err
-               }
+       mergePatch, err := wfc.getPodFinalizerRemovalPatch(namespace, podName)
+       if err != nil {
+               return err
+       }
+       if mergePatch == nil {
+               return nil
+       }
+
+       // merge patch is idempotent
+       if _, err := wfc.kubeclientset.CoreV1().Pods(namespace).Patch(
+               ctx,
+               podName,
+               types.MergePatchType,
+               mergePatch,
+               metav1.PatchOptions{},
+       ); err != nil {
+               return err
        }
        return nil
 }
@@ -619,18 +639,20 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
                                return err
                        }
                case labelPodCompleted:
-                       patch := `{"metadata": {"labels": {"workflows.argoproj.io/completed": "true"}}}`
-                       if finalizers, found := wfc.getPodFinalizerRemovalPatch(namespace, podName); found {
-                               patch = fmt.Sprintf(`{"metadata": {"finalizers": %s, "labels": {"workflows.argoproj.io/completed": "true"}}}`, finalizers)
+                       mergePath, err := wfc.getPodFinalizerRemovalPatch(namespace, podName)
+                       if err != nil {
+                               return err
                        }
-                       _, err := pods.Patch(
+                       if mergePath == nil {
+                               return nil
+                       }
+                       if _, err := pods.Patch(
                                ctx,
                                podName,
                                types.MergePatchType,
-                               []byte(patch),
+                               mergePath,
                                metav1.PatchOptions{},
-                       )
-                       if err != nil {
+                       ); err != nil {
                                return err
                        }
                case deletePod:
```

Reply:
I can write a new function `getPodCleanupPatch(namespace, podName string, labelPodCompleted, removeFinalizer bool) ([]byte, erroir) {}` to generate the patch. About if using the json-patch, I think the original way is more straight-forward, in our environment, some pod's size can be greater than hundred KBs, which could cost extra cpu and memory if using json-patch.

Reply:
> I think the original way is more straight-forward, in our environment, some pod's size can be greater than hundred KBs, which could cost extra cpu and memory if using json-patch.

I see. Your point makes perfect sense to me.

> I can write a new function getPodCleanupPatch(namespace, podName string, labelPodCompleted, removeFinalizer bool) ([]byte, erroir) {} to generate the patch.

It might be more efficient than handling strings.